File: server/src/SwipeJobs.Application/Common/PipelineArchitecture.cs
using SwipeJobs.Domain.Enums;

namespace SwipeJobs.Application.Common;

/// <summary>
/// Canonical hiring pipeline model for employer UX. Kanban UI maps to these columns.
/// </summary>
public static class PipelineArchitecture
{
    public enum Column
    {
        Applied,
        Reviewing,
        Shortlisted,
        Interview,
        Offer,
        Hired,
        Rejected,
    }

    public static Column ResolveColumn(ApplicationStatus status, InterviewPhase interviewPhase)
    {
        return status switch
        {
            ApplicationStatus.Pending or ApplicationStatus.Applied => Column.Applied,
            ApplicationStatus.UnderReview => Column.Reviewing,
            ApplicationStatus.Shortlisted => Column.Shortlisted,
            ApplicationStatus.InterviewInvited or ApplicationStatus.Interviewing => Column.Interview,
            ApplicationStatus.OfferSent => Column.Offer,
            ApplicationStatus.Hired => Column.Hired,
            ApplicationStatus.Rejected or ApplicationStatus.Withdrawn => Column.Rejected,
            _ => Column.Applied,
        };
    }

    /// <summary>
    /// Interview column uses InterviewPhase for requested / scheduled / completed before offer.
    /// </summary>
    public static InterviewPhase ResolveInterviewPhase(ApplicationStatus status, InterviewPhase storedPhase)
    {
        if (status is not (ApplicationStatus.InterviewInvited or ApplicationStatus.Interviewing))
            return InterviewPhase.None;

        return storedPhase switch
        {
            InterviewPhase.Scheduled or InterviewPhase.Completed or InterviewPhase.Requested => storedPhase,
            _ => status == ApplicationStatus.InterviewInvited
                ? InterviewPhase.Requested
                : InterviewPhase.None,
        };
    }

    public static IReadOnlyList<Column> OrderedColumns { get; } =
    [
        Column.Applied,
        Column.Reviewing,
        Column.Shortlisted,
        Column.Interview,
        Column.Offer,
        Column.Hired,
        Column.Rejected,
    ];
}

File: client/src/lib/employer/pipelineArchitecture.ts
import { ApplicationStatus } from '@/models/enums';

/** Kanban column identifiers — canonical employer pipeline model. */
export type PipelineColumnId =
  | 'applied'
  | 'reviewing'
  | 'shortlisted'
  | 'interview'
  | 'offer'
  | 'hired'
  | 'rejected';

/** Sub-state within the Interview column (calendar/scheduling ready). */
export type InterviewPhase = 'none' | 'requested' | 'scheduled' | 'completed';

export interface PipelineColumnDefinition {
  id: PipelineColumnId;
  label: string;
  /** Application statuses grouped into this column */
  statuses: ApplicationStatus[];
  /** Interview sub-phases shown inside the Interview column */
  interviewPhases?: InterviewPhase[];
}

export const PIPELINE_COLUMNS: PipelineColumnDefinition[] = [
  {
    id: 'applied',
    label: 'Applied',
    statuses: [ApplicationStatus.Pending, ApplicationStatus.Applied],
  },
  {
    id: 'reviewing',
    label: 'Reviewing',
    statuses: [ApplicationStatus.UnderReview],
  },
  {
    id: 'shortlisted',
    label: 'Shortlisted',
    statuses: [ApplicationStatus.Shortlisted],
  },
  {
    id: 'interview',
    label: 'Interview',
    statuses: [ApplicationStatus.InterviewInvited, ApplicationStatus.Interviewing],
    interviewPhases: ['requested', 'scheduled', 'completed'],
  },
  {
    id: 'offer',
    label: 'Offer',
    statuses: [ApplicationStatus.OfferSent],
  },
  {
    id: 'hired',
    label: 'Hired',
    statuses: [ApplicationStatus.Hired],
  },
  {
    id: 'rejected',
    label: 'Rejected',
    statuses: [ApplicationStatus.Rejected, ApplicationStatus.Withdrawn],
  },
];

export const INTERVIEW_PHASE_LABELS: Record<InterviewPhase, string> = {
  none: 'Interview',
  requested: 'Interview requested',
  scheduled: 'Interview scheduled',
  completed: 'Interview completed',
};

export function resolvePipelineColumn(status: ApplicationStatus): PipelineColumnId {
  const match = PIPELINE_COLUMNS.find((column) => column.statuses.includes(status));
  return match?.id ?? 'applied';
}
